### DISCUSSION TITLE
[Curious] Do-one-thing-do-it-well vs Monolith Framework & Wildcard API vs No-API

### DISCUSSION THREAD

**Discussion by brillout** [blitz-js/blitz#736] [General]
Hi @flybayer 

Nice job on creating Blitz! 

I'm curious about what you think about https://github.com/reframejs/reframe. The idea is to have starters + libs instead of one big monolithic framework.

Which brings me to https://github.com/reframejs/wildcard-api.

I'm the author of both.

I really like your manifest, that was my goal with Reframe as well until I got caught up by https://lsos.org/.

Thanks for what you are doing and if by any chance Blitz eventually follows what I call the "zero-lock framework" principle I'd your biggest advocate :-).

**Comment by flybayer** [comment]
Hey @brillout! I had never heard of Reframe or Wildcard. Wildcard is really cool! Definitely similar to what we are doing with Blitz.

I think there's definitely a place for a starters + libs framework, but for Blitz we specifically want it to be a monolith like Rails. There's a fundamental upper limit to the DX of starters+libs that we are going beyond :) 

Rails is very popular for this reason. The main reason Rails is falling out of favor is because people want to build SPA type experiences with fullstack Javascript and with ability to share code with React Native apps. So Blitz is filling that hole :)

**Comment by brillout** [comment]
> fundamental upper limit to the DX of starters+libs 

Are you sure about this?

To me, this is actually the opposite. It's nice (and reassuring) to use do-one-thing libraries highly focused and well maintained instead of using one big monolith that will much more likely die. I believe it's not only me but the general sentiment.

And, even more importantly, I can much easier replace a do-one-thing lib with another one. Replacing a framework is a order of magnitude more daunting task.

Curious about your thoughts about this.

Getting this right is fundamentally important for Blitz.

**Reply by grantlonie** [reply]
Tend to agree here. I get nervous when libs try to do too much 😬

**Comment by brillout** [comment]
@merelinguist You downvoted; do you mind sharing your thoughts? I'm genuinely curious

**Comment by raphaelbadia** [comment]
Rails, (or Laravel in php) are frameworks that follow the “convention over configuration” paradigm. 
In my experience, do-one-thing libs are mostly liked by senior developers that want full control over their apps and know how to deal with the things they might break.
A monolith helps beginners follow “the path”, even if it’s not the best it helps them get going faster.
For intermediate, it’s still convenient to jump on a bulletproof architecture designed by developers 4x more experienced than them.

**Reply by brillout** [reply]
I agree, that said a starters + libs framework could guide users as well as a monolithic framework.

**Comment by merelinguist** [comment]
@brillout We do see the value in extracting certain things out as necessary. Yesterday, we published `Superjson`, which we want to use within Blitz for serialization. We’ve collaborated with Gatsby on our recipes feature, and will likely work with them on publishing some common-ground code. I believe Brandon has spoken to the Redwood team about collaborating on a shared auth package.

That said, there’s both a danger and a difficulty in splitting everything up. The difficulty is that Blitz is not that big right now! I’d prefer to give our users a complete package, rather than a dozen little ones. The danger is that, at least at our scale, it’s a lot easier to maintain a monolith than too many “microservices”. When everything is in the same place, and uses the same mental model, it’s generally easier to maintain.

A great example of a library which initially started with the principle of splitting things up is Apollo. A few years ago, you had to install a whole list of packages to get Apollo working, a few of which you could swap as you wished. Over time, Apollo seems to have slowly consolidated these because it’s much easier to install one package and go, than to install a handful, many of which have seemingly opaque functionality.

The final thing I’d say is that Blitz is a framework. Fundamentally, you are going to have to buy in to some of the design decisions Blitz at some level, otherwise there’s not really any framework, but just a collection of utility libraries! I think Blitz is a much stronger framework for recommending what we feel is the best way to do things, rather than leaving users fending for themselves.

Interested to hear your thoughts.

**Comment by brillout** [comment]
Sounds good!

The more code lives in libraries the better! If 50% of the code you guys write are for libraries that can be used independently of Blitz, then you are doing Blitz and the community as a whole a huge favor.

Libraries are inherently more robust and flexible than monoliths.

Where can I find the auth package you mentioned? I was actually looking for one for Reframe but I never found something decent. So I'm super curious about what you are building there.

> We published Superjson, which we want to use within Blitz for serialization.

Neat, I may actually use Superjson instead of the serializer I wrote for Wildcard. That's really neat that instead of competing we end up sharing code. Libraries enable fundamentally more collaboration :).

I've gathered an extensive amount of feedback and experience with Wildcard. Maybe we should cooperate on the RPC side &mdash; is that something you guys would be up to? I'm working on a Wildcard rewrite and it's going to be much much better. New name, new docs website, simpler context object, etc.
